### PR TITLE
Add binary index support for Lucene engine

### LINKED ISSUE #1857: [FEATURE]Binary vector support with Lucene
Similar to https://github.com/opensearch-project/k-NN/issues/1764, I like to see binary vector support with Lucene engine

### PR DESCRIPTION
**Pull request by jed326**
### Description
Add binary vector support for Lucene engine. For testing I parameterized `BinaryIndexIT.java` to run with both `FAISS` and `LUCENE` engines.

### Related Issues
Resolves #1857

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).
    - [x] https://github.com/opensearch-project/documentation-website/issues/8955

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jed326** [comment]
@heemin32 @navneet1v could you please help review? Also it looks like the linux build is broken.

**Comment by navneet1v** [comment]
@peterzhuamazon are there changes in the CI image? we are seeing failures due to GlibC

**Comment by peterzhuamazon** [comment]
> @peterzhuamazon are there changes in the CI image? we are seeing failures due to GlibC

Hi @navneet1v you can follow the guide here to mitigate the issue:
* https://github.com/opensearch-project/opensearch-build/issues/5178

Please let me know if you have more questions, thanks!

**Comment by navneet1v** [comment]
@jed326 I raised a PR for the fix: https://github.com/opensearch-project/k-NN/pull/2303

**Comment by navneet1v** [comment]
@jed326 please pull in latest changes from main branch. CI setup fix is done.

File: src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990HnswBinaryVectorsFormat.java
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990HnswBinaryVectorsFormat.java:84]: This value has been changed for K-NN plugin. For k-NN plugin Max dimension is 16k. 
Ref: https://github.com/opensearch-project/k-NN/blob/009c8f39d541f3edb2d665e4300257bee0d97d83/src/main/java/org/opensearch/knn/index/engine/KNNEngine.java#L37-L45

File: src/main/java/org/opensearch/knn/index/codec/params/KNNVectorsFormatParams.java
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/codec/params/KNNVectorsFormatParams.java:31]: do we need both of these constructors? can we not just keep 1.

File: src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java:66]: I thought hamming distance is all about counting bits by doing xor. why do we need dimensions here?

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java:66]: and by looking at this formula I think we just need to do this:
```
1 / (1+VectorUtil.xorBitCount(queryVector, vectorValues.vectorValue(node)))
```

https://opensearch.org/docs/latest/search-plugins/knn/approximate-knn/#spaces

File: src/main/java/org/opensearch/knn/index/VectorDataType.java
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/VectorDataType.java:43]: instead of 8 use `Byte.SIZE`

File: src/main/java/org/opensearch/knn/index/KNNVectorSimilarityFunction.java
Comment (jed326) [reply src/main/java/org/opensearch/knn/index/KNNVectorSimilarityFunction.java:33]: I agree that this is not ideal, but at a high level this seems to be how Lucene expects it as we can pass in the custom scorer to ignore this.

At a lower level, this method is invoked here: https://github.com/opensearch-project/k-NN/blob/270ac6a52f2ff9cc9cfa58dbd65333b21f925048/src/main/java/org/opensearch/knn/index/mapper/LuceneFieldMapper.java#L103-L107

And the `KnnByteVectorField` requires a vector similarity function to be passed in to it even if it is not being used:

https://github.com/opensearch-project/k-NN/blob/88792e42f121b050f2fc9cf32b039052aab62128/src/main/java/org/opensearch/knn/index/VectorDataType.java#L72

So it seems like we need to pass in a `VectorSimilarityFunction` regardless, wdyt?

File: src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990HnswBinaryVectorsFormat.java
Comment (jed326) [reply src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990HnswBinaryVectorsFormat.java:22]: I also wasn't sure on this. The `KNN9120Codec` codec maps to the `KNN990PerFieldKnnVectorsFormat` class so I felt `KNN990HnswBinaryVectorsFormat` was more in line with what was present.

File: src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java
Comment (jed326) [reply src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java:66]: Good question. I was referencing [`FlatBitVectorsScorer#score`](https://github.com/apache/lucene/blob/d8a81476262312359a67dee9c6c43e2a527bd6e5/lucene/codecs/src/java/org/apache/lucene/codecs/bitvectors/FlatBitVectorsScorer.java#L77-L81) and based on the formula it looks like the dimension is used to normalize the score. I don't think we are doing that normalization for the faiss engine so we can probably skip doing that here too.

File: src/main/java/org/opensearch/knn/index/codec/params/KNNVectorsFormatParams.java
Comment (jed326) [reply src/main/java/org/opensearch/knn/index/codec/params/KNNVectorsFormatParams.java:31]: I didn't want to modify all the places the first constructor was being used so I'm basically overloading it like so.

**Comment by jed326** [comment]
@navneet1v thanks for the feedback, mind taking another look after the latest revisions?

File: src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java]: Yes lets skip that normalization for now. But did you find any reason in which cases the normalization will be helpful?

File: src/main/java/org/opensearch/knn/index/codec/params/KNNVectorsFormatParams.java
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/codec/params/KNNVectorsFormatParams.java:31]: Yeah I understand. I will less code here by just having 1 constructor, and the old constructor I feel would be just getting used in tests so I feel it is safe to change the code there. 

But since this is not a make or break feature, I will leave this upto your best judgement.

File: src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990HnswBinaryVectorsFormat.java
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990HnswBinaryVectorsFormat.java]: @jed326, `KNN990PerFieldKnnVectorsFormat` is for older version of lucene. I think since you are making changes in the format code too you should create a `KNN9120PerFieldKnnVectorsFormat` class and then `KNN9120HnswBinaryVectorsFormat` for consistency. @jmazanec15 please comment if you think otherwise.

File: src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java
Comment (jed326) [reply src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNN990BinaryVectorScorer.java]: I was not able to find any cases which the normalization is needed in either the Lucene discussions or in FAISS documentation.

File: src/main/java/org/opensearch/knn/index/KNNVectorSimilarityFunction.java
Comment (jed326) [reply src/main/java/org/opensearch/knn/index/KNNVectorSimilarityFunction.java:33]: Thanks @navneet1v that sounds reasonable to me so we can keep `KNNVectorSimilarityFunction` as essentially the source of truth for mapping the space type to the Lucene internal similarity function.

**Comment by jed326** [comment]
@navneet1v added `KNN9120PerFieldKnnVectorsFormat`, etc in the latest revision. Thanks!

File: src/main/java/org/opensearch/knn/index/codec/KNN9120Codec/KNN9120BinaryVectorScorer.java
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/codec/KNN9120Codec/KNN9120BinaryVectorScorer.java:65]: Lets use `KNNVectorSimilarityFunction.HAMMING.compare()` function here to keep the Hamming distance logic consistent.

**Review by navneet1v — APPROVED** [review]
Overall looks good to me. Just 1 minor comments

File: src/main/java/org/opensearch/knn/index/codec/KNN9120Codec/KNN9120BinaryVectorScorer.java
Comment (jed326) [reply src/main/java/org/opensearch/knn/index/codec/KNN9120Codec/KNN9120BinaryVectorScorer.java:65]: Good call -- that takes care of the equal dimension checks as well, thanks!

**Comment by jed326** [comment]
@heemin32 @jmazanec15 could one of you help with a second review here? Thanks!

**Comment by heemin32** [comment]
LGTM. Could you also add integ test with lucene engine in `FilteredSearchBinaryIT` and `NestedSearchBinaryIT`?

**Comment by jed326** [comment]
> LGTM. Could you also add integ test with lucene engine in `FilteredSearchBinaryIT` and `NestedSearchBinaryIT`?

Thanks @heemin32 -- added the same engine parameterization to these two classes as well in the latest revision.

**Review by heemin32 — APPROVED** [review]

**Review by navneet1v — APPROVED** [review]

**Comment by navneet1v** [comment]
Lets wait for all CIs to complete before we can merge this change.

@jed326 lets complete the benchmarks for Lucene binary vectors, @heemin32 can you please share details on how benchmarks was done for this blog: https://opensearch.org/blog/lower-your-cost-on-opensearch-using-binary-vectors/

and lets kick off the documentation update too for this feature.

**Comment by jed326** [comment]
Created the docs issue here: https://github.com/opensearch-project/documentation-website/issues/8955

**Pull request merged**
